### ISSUE TITLE
DL3008: How to pin versions in apt get install without breaking future builds?

### ISSUE THREAD

**Issue by Delgan** [hadolint/hadolint#884]
Hi.

The [DL3008](https://github.com/hadolint/hadolint/wiki/DL3008) warning recommends to pin versions in Dockerfile.

This makes a lot of sense. However, since I'm following this advice, I'm repeatedly facing the same problem: the required versions are no longer available after a few weeks. 

This Dockerfile used to work fine at the time I created it:

```Dockerfile
FROM ubuntu:22.04

ARG USERNAME=dev
ARG USER_UID=1000
ARG USER_GID=$USER_UID

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get -y update \
    && apt-get install -y --no-install-recommends \
        sudo=1.9.9-1ubuntu2 \
        apt-utils=2.4.7 \
        git=1:2.34.1-1ubuntu1.4 \
        wget=1.21.2-2ubuntu1 \
        vim=2:8.2.3995-1ubuntu2 \
        ssh=1:8.9p1-3 \
        xz-utils=5.2.5-2ubuntu1 \
        libasound2=1.2.6.1-1ubuntu1 \
        libnss3=2:3.68.2-0ubuntu1.1 \
        libatk1.0-0=2.36.0-3build1 \
        libatk-bridge2.0-0=2.38.0-3 \
        libgtk-3.0=3.24.33-1ubuntu2 \
        xvfb=2:21.1.3-2ubuntu2.1 \
        python3-pip=22.0.2+dfsg-1 \
        shellcheck=0.8.0-2 \
        bash-completion=1:2.11-5ubuntu1 \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*
```

However, trying to build the image today fails with:

```bash
E: Version '1:2.37.2-0ppa1~ubuntu20.04.1' for 'git' was not found
E: Version '2:8.1.2269-1ubuntu5.7' for 'vim' was not found
```

I understand it's failing because the distribution updates the available packages and remove the old ones, but do you know what is be best practice to make sure the Docker image can be more or less rebuild anytime?

**Comment by m-ildefons** [comment]
Hi @Delgan ,
Version pinning and consistent rebuilds are mutually exclusive. You can of course use your own PPA to host the packages in the versions that you pinned, then it would work every time, but I would understand if that is too much. Or you can build your own base image with your dependencies included and always build from that...

The rationale behind version pinning is that it gives _consistent_ builds i.e. the resulting artifact doesn't change and (assuming you can still find the packages in the repos) the build doesn't fail because of changing library APIs or such. Most package managers also support version constraints (e.g. vim>=8.0.0) to avoid that problem until a major version change occurs.

On the other hand just using whatever is in the repo can be nice, since the package manager will always find a version to install and usually, this also results in up to date packages, but as I said, it's hard to ensure that the build and runtime dependencies don't break their APIs this way and it's also hard to trace back what artifacts actually went into the build.

In the end whether you use version pinning or not can be considered a matter of taste or at least an engineering compromise as there are valid reasons for and against it. I haven't seen a solution that provides both benefits without any disadvantages.
Also Hadolint is just suggesting an opinionated version of best practices and if they don't align with your project's goals, it's perfectly fine to disable certain warnings or give them a lower priority. Check our documentation on how to do that.

Hope this answers your question...

**Comment by Delgan** [comment]
Thanks a lot @m-ildefons for taking the time to answer me in such detail!

It definitely clarifies my concerns. I was surprised at first, and I wondered if I was doing something incorrectly, but it confirms that version pinning is not easily compatible with all use cases.

It allows me to see two different practices:
* If this is the built `image` which is distributed to "end user", then it makes sense to use versions pinning. The `image` likely won't change very often and the `Dockerfile` serves as a reference. The package versions are documented and it ensure some reproducibility inside the container for whoever uses the image.
* If this is the raw `Dockerfile` which is distributed to "end user", then versions pinning makes less sense because we have no control about when the image will be built. The versions will likely become unavailable, preventing the user to build the image himself. In such a case, it's OK to disable DL3008 for the non-critical packages.

That summarizes my thoughts on this, thanks!
